### PR TITLE
Namespace pool limiter feature

### PR DESCRIPTION
**Pull request by maknop**

### CODE REVIEW

File: controllers/cloud.redhat.com/pool_controller.go [hunk 1]
```diff
@@ -76,19 +76,21 @@ func (r *NamespacePoolReconciler) Reconcile(ctx context.Context, req ctrl.Reques
...
 
-	quantityOfNamespaces := r.checkReadyNamespaceQuantity(pool)
+	quantityOfNamespaces := r.checkNamespaceQuantity(pool, statusTypeCount)
 
 	if quantityOfNamespaces > 0 {
 		r.Log.Info(fmt.Sprintf("Filling '%s' pool with %d namespace(s)", pool.Name, quantityOfNamespaces))
 		err := r.increaseReadyNamespacesQueue(ctx, pool, quantityOfNamespaces)
 		if err != nil {
 			r.Log.Error(err, fmt.Sprintf("unable to create more namespaces for '%s' pool.", pool.Name))
+			return ctrl.Result{Requeue: true}, err
```
Comment (psav) [inline controllers/cloud.redhat.com/pool_controller.go:86]: Good catch

File: controllers/cloud.redhat.com/pool_controller.go [hunk 2]
```diff
@@ -135,6 +137,7 @@ func (r *NamespacePoolReconciler) getPoolStatus(ctx context.Context, pool crd.Na
 
 	var readyNS int
 	var creatingNS int
+	var totalNS int
```
Comment (psav) [inline controllers/cloud.redhat.com/pool_controller.go:140]: I'd argue this isn't necessary here, we already know the total by adding creating and ready

File: controllers/cloud.redhat.com/pool_controller.go [hunk 3]
```diff
@@ -149,25 +152,38 @@ func (r *NamespacePoolReconciler) getPoolStatus(ctx context.Context, pool crd.Na
...
 }
 
-func (r *NamespacePoolReconciler) checkReadyNamespaceQuantity(pool crd.NamespacePool) int {
+func (r *NamespacePoolReconciler) checkNamespaceQuantity(pool crd.NamespacePool, statusTypeCount map[string]int) int {
 	size := pool.Spec.Size
+	limiter := pool.Spec.SizeLimiter
 	ready := pool.Status.Ready
 	creating := pool.Status.Creating
 
+	if limiter > 0 {
```
Comment (psav) [inline controllers/cloud.redhat.com/pool_controller.go:175]: Combine the ifs here, single line with an `&&`

File: controllers/cloud.redhat.com/pool_controller.go [hunk 4]
```diff
@@ -149,25 +152,38 @@ func (r *NamespacePoolReconciler) getPoolStatus(ctx context.Context, pool crd.Na
...
+		if statusTypeCount["total"] >= limiter {
+			r.Log.Info(fmt.Sprintf("Max number of namespaces for pool [%s] already created or in-use. Max of [%d] allowed.", pool.Name, limiter))
```
Comment (psav) [inline controllers/cloud.redhat.com/pool_controller.go:177]: So the problem here is that it only takes into account the number that are in the pool and the number being created, which actually should be equal to `size`. (The only time it's not if is the pool has a size `x` and the pool size is reduced by `y`, but that requires someone to make changes to the spec)

Actually what we really want here is a limit on the TOTAL number of namespaces that have come from that pool. So we should look at namespaces which are also checked out :)

File: controllers/cloud.redhat.com/pool_controller.go (hunk 2 above)
Comment (maknop) [reply controllers/cloud.redhat.com/pool_controller.go:140]: "ready" only records namespaces ready to be reserved. It checks the owner reference's UID against the pool's UID:
`
for _, owner := range ns.GetOwnerReferences() {
	if owner.UID == pool.GetUID() {
		switch ns.Annotations["env-status"] {
		case "ready":
			readyNS++
		case "creating":
			creatingNS++
		case "error":
			r.Log.Info("Error status for namespace. Prepping for deletion.", "ns-name", ns.Name)
			errNamespaceList = append(errNamespaceList, ns.Name)
		}
	}
}
`

**Pull request closed without merging**